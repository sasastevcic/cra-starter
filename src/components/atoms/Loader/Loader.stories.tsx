import { ComponentStory, ComponentMeta } from '@storybook/react';
import LoaderComponent from './index';

export default {
	title: 'UI Elements/Loader',
	component: LoaderComponent,
} as ComponentMeta<typeof LoaderComponent>;

const Template: ComponentStory<typeof LoaderComponent> = (args) => <LoaderComponent {...args} />;

export const Loader = Template.bind({});
Loader.parameters = {
	backgrounds: { default: 'light' },
};
