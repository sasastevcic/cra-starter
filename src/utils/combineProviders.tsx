/* eslint-disable @typescript-eslint/no-explicit-any */
import { ComponentProps, ComponentType, FunctionComponent, ReactNode } from 'react';

type Providers = Array<[ComponentType<any>, ComponentProps<any>?]>;

export const combineProviders = (
	providers: Providers,
): FunctionComponent<{ children: ReactNode }> =>
	providers.reduce(
		(AccumulatedProviders, [Provider, props = {}]) =>
			({ children }) =>
				(
					<AccumulatedProviders>
						<Provider {...props}>{children}</Provider>
					</AccumulatedProviders>
				),
		({ children }) => <>{children}</>,
	);
