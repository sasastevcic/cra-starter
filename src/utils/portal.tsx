import { ReactNode, ReactPortal, useMemo } from 'react';
import { createPortal } from 'react-dom';

interface Portal {
	portalId: string;
	children: ReactNode;
}

const Portal = ({ portalId, children }: Portal): ReactPortal => {
	const portalElement = useMemo(() => document.getElementById(portalId), [portalId]);

	if (!portalElement) {
		throw new Error(`Element not found! Element id: ${portalId}`);
	}

	return createPortal(children, portalElement);
};

export default Portal;
