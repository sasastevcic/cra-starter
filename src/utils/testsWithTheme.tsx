/* eslint-disable @typescript-eslint/ban-ts-comment */
import { ReactElement, ComponentType, ReactNode } from 'react';
import { render, RenderOptions, RenderResult } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import { theme } from '../styles/config/theme';

export const componentWithTheme = (component: ReactNode): ReactElement => (
	// @ts-ignore
	<ThemeProvider theme={theme}>{component}</ThemeProvider>
);

export const renderWithTheme = (component: ReactElement, options?: RenderOptions): RenderResult => {
	const wrapper: ComponentType<{ children: ReactNode }> = ({ children }) =>
		componentWithTheme(children);

	return render(component, { wrapper, ...options });
};
