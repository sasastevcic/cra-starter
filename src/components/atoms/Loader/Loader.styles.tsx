import styled from 'styled-components';
import { keyframes } from '../../../styles/config/keyframes';
import { TransitionEase } from '../../../styles/config/variables';
import { OUTLINE_SIZE, RADIUS, SIZE, STROKE_WIDTH } from './Loader.config';

export const StyledLoaderHolder = styled.div`
	width: 100%;
	text-align: center;
	padding: 1rem 0;
`;

export const StyledLoader = styled.svg.attrs(() => ({
	width: SIZE,
	height: SIZE,
}))`
	animation: ${keyframes.rotate} 2s linear infinite;
`;

const StyledCircle = styled.circle.attrs(() => ({
	cx: SIZE / 2,
	cy: SIZE / 2,
	r: RADIUS,
}))`
	stroke-width: ${STROKE_WIDTH}px;
	fill: none;
`;

export const StyledBaseCircle = styled(StyledCircle)`
	stroke: ${({ theme }) => theme.color.whiteFaded};
`;

export const StyledAnimatingCircle = styled(StyledCircle)`
	stroke-dasharray: 50, ${OUTLINE_SIZE};
	stroke-dashoffset: 0;
	animation: ${keyframes.dash} 1.5s ${TransitionEase.Default} infinite;
	stroke: ${({ theme }) => theme.color.black};
`;
