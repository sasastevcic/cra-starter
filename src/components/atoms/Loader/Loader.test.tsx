import renderer from 'react-test-renderer';
import { renderWithTheme, componentWithTheme } from '../../../utils/testsWithTheme';
import Loader from './index';

describe('Loader', () => {
	it('should render', () => {
		const { getByTestId } = renderWithTheme(<Loader />);

		const element = getByTestId('Loader');

		expect(element).toBeInTheDocument();
	});

	it('should render correctly', () => {
		const tree = renderer.create(componentWithTheme(<Loader />)).toJSON();

		expect(tree).toMatchSnapshot();
	});
});
