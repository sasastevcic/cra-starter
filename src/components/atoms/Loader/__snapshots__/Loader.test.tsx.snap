// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loader should render correctly 1`] = `
<div
  className="sc-bczRLJ dVNMRw"
  data-testid="Loader"
>
  <svg
    className="sc-gsnTZi Dpegn"
    height={60}
    width={60}
  >
    <circle
      className="sc-dkzDqf sc-hKMtZM exNUlu gwUwEt"
      cx={30}
      cy={30}
      r={18}
    />
    <circle
      className="sc-dkzDqf sc-eCYdqJ exNUlu dcvWlS"
      cx={30}
      cy={30}
      r={18}
    />
  </svg>
</div>
`;
